src/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/Base64Transforms.cs

```diff
@@ -96,143 +112,226 @@ public class FromBase64Transform : ICryptoTransform
 
-            byte[] transformBuffer = new byte[_inputIndex + effectiveCount];
-            Buffer.BlockCopy(_inputBuffer, 0, transformBuffer, 0, _inputIndex);
-            Buffer.BlockCopy(temp, 0, transformBuffer, _inputIndex, effectiveCount);
+            for (int i = 0; i < inputBuffer.Length; i++)
+            {
+                if (!IsWhitespace(inputBuffer[i]))
+                {
+                    tmpBuffer[count++] = inputBuffer[i];
+                }
+            }
 
-            _inputIndex = (effectiveCount + _inputIndex) % 4;
-            Buffer.BlockCopy(temp, effectiveCount - _inputIndex, _inputBuffer, 0, _inputIndex);
+            return tmpBuffer.Slice(0, count);
+        }
 
-            char[] tempChar = Encoding.ASCII.GetChars(transformBuffer, 0, 4 * numBlocks);
-            byte[] tempBytes = Convert.FromBase64CharArray(tempChar, 0, 4 * numBlocks);
-            return tempBytes;
+        private static bool IsWhitespace(byte value)
+        {
+            // We assume ASCII encoded data. If there is any non-ASCII char, it is invalid
+            // Base64 and will be caught during decoding.
+
+            // SPACE        32
+            // TAB           9
+            // LF           10
+            // VTAB         11
+            // FORM FEED    12
+            // CR           13
+
+            // RyuJIT produces better code with the ternary. This JIT-bug is tracked in https://github.com/dotnet/coreclr/issues/914
+            return value == 32 || ((uint)value - 9 <= (13 - 9));
         }
 
-        private byte[] DiscardWhiteSpaces(byte[] inputBuffer, int inputOffset, int inputCount)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetOutputSize(int bytesToTransform, Span<byte> tmpBuffer)
         {
-            int i, iCount = 0;
-            for (i = 0; i < inputCount; i++)
+            int outputSize = Base64.GetMaxDecodedFromUtf8Length(bytesToTransform);
+
+            const byte padding = (byte)'=';
+            int len = tmpBuffer.Length;
+
+            // In Base64 there are maximum 2 padding chars
+
+            if (tmpBuffer[len - 2] == padding)
             {
-                if (char.IsWhiteSpace((char)inputBuffer[inputOffset + i])) iCount++;
+                outputSize--;
             }
 
-            // If there's nothing to do, leave early
-            if (iCount == 0 && inputOffset == 0 &&
-                inputCount == inputBuffer.Length)
+            if (tmpBuffer[len - 1] == padding)
             {
-                return inputBuffer;
+                outputSize--;
             }
 
-            byte[] rgbOut = new byte[inputCount - iCount];
-            iCount = 0;
-            for (i = 0; i < inputCount; i++)
+            return outputSize;
+        }
+
+        private void ConvertFromBase64(Span<byte> tmpBuffer, Span<byte> outputBuffer, out int consumed, out int written)
+        {
+            int bytesToTransform = _inputIndex + tmpBuffer.Length;
+
```

Comment:
Might be good to assert that it's 4 or more.  I think that any number less than 3 where _inputIndex is non-zero will make the "save this data" line throw, but it helps document-via-code that this is only expected to be used once output will be produced.